src/main/scala/com/databricks/spark/avro/ConfluentSparkAvroUtils.scala

```diff
@@ -60,12 +66,59 @@ class ConfluentSparkAvroUtils(schemaRegistryURLs: String) extends Serializable {
 
   @transient private lazy val deserializer = new KafkaAvroDeserializer(schemaRegistry)
 
+  @transient private lazy val AES_MAGIC_BYTES = Set[Byte](2, 3)
+
+  @transient private lazy val awsKms: AWSKMS = new AWSKMSClient()
+
+  @transient private lazy val keyCache = MutableMap[Seq[Byte], Array[Byte]]()
+
+  val HEADER_SIZE = 5
+  val AES_ENCRYPTED_KEY_SIZE = 184
+
   def deserializerForSubject(subject: String, version: String): UserDefinedFunction = {
     udf(
       (payload: Array[Byte]) => {
+        val magic_byte: Byte = payload(0)
+
+        var data: Array[Byte] =
+          if (AES_MAGIC_BYTES(magic_byte)) {
+            // Decrypt Key
+            val encryptedKey: Array[Byte] = payload.slice(HEADER_SIZE,
+                                                          HEADER_SIZE + AES_ENCRYPTED_KEY_SIZE)
+            var plainKeyBytes: Array[Byte] =
+              if (keyCache.keySet.contains(encryptedKey))
```

Comment:
`keyCache.getOrElseUpdate(encrypedKey, compute(encrypedKey))` where `compute()` is the function I am suggesting you to define in the next comment.